spanner/api/SpannerTest/Tests.cs

```diff
@@ -217,22 +217,23 @@ async void InitializeInstance()
                 string name = $"projects/{_fixture.ProjectId}/instances/{_fixture.InstanceId}";
                 Instance response = await instanceAdminClient.GetInstanceAsync(name);
             }
-            catch (RpcException ex) when (ex.Status.StatusCode == StatusCode.NotFound)
+            catch (AggregateException e)
             {
-                string parent = $"projects/{_fixture.ProjectId}";
-                Instance instance = new Instance
+                bool rethrow = true;
+                foreach (var innerException in e.InnerExceptions)
                 {
-                    DisplayName = _fixture.InstanceId,
-                    Name = $"projects/{_fixture.ProjectId}/instances/{_fixture.InstanceId}",
-                    NodeCount = 1,
-                    Config = $"projects/{_fixture.ProjectId}/instanceConfigs/regional-us-central1"
-                };
-
-                // Make the CreateInstance request
-                var response = instanceAdminClient.CreateInstance(parent, _fixture.InstanceId, instance);
-
-                // Poll until the returned long-running operation is complete
-                response.PollUntilCompleted();
+                    SpannerException spannerException = innerException as SpannerException;
+                    if (spannerException != null && spannerException.ErrorCode == ErrorCode.NotFound)
+                    {
+                        _spannerCmd.Run("createInstance", _fixture.ProjectId, _fixture.InstanceId);
+                        rethrow = false;
+                        break;
+                    }
+                }
+                if (rethrow)
+                {
+                    throw;
+                }

```

Comment:
@laljikanjareeya catching this exception doesn't work unfortunately. I'm not sure I understand why.

The error printed out looks like this so I'm a bit confused by why it's not caught correctly.

```
  Error Message:
   System.AggregateException : One or more errors occurred. (Object not found.)
---- Google.Cloud.Spanner.Data.SpannerException : Object not found.
-------- Grpc.Core.RpcException : Status(StatusCode=NotFound, Detail="Instance not found: projects/appdev-soda-spanner-staging/instances/test-create-instance")
  Stack Trace:
     at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at GoogleCloudSamples.Spanner.Program.CreateSampleDatabase(String projectId, String instanceId, String databaseId) in /google-cloud-dotnet/spanner/api/Spanner/Program.cs:line 1929
   at GoogleCloudSamples.Spanner.Program.<>c.<Main>b__124_1(CreateSampleDatabaseOptions opts) in /google-cloud-dotnet/spanner/api/Spanner/Program.cs:line 3689
```

Reply:
@skuruppu `instanceAdminClient.GetInstanceAsync` will throw `RpcException` not `AggregateException`  because we are not using command line for getting instance, so below code should work fine.

```csharp
async void InitializeInstance()
{
    InstanceAdminClient instanceAdminClient = await InstanceAdminClient.CreateAsync();
    string name = $"projects/{_fixture.ProjectId}/instances/{_fixture.InstanceId}";
    try
    {
        Instance response = await instanceAdminClient.GetInstanceAsync(name);
    }
    catch (RpcException ex) when (ex.Status.StatusCode == StatusCode.NotFound)
    {
        _spannerCmd.Run("createInstance", _fixture.ProjectId, _fixture.InstanceId);
    }
}
```

Reply:
A couple of things happening here:

1. If you look closely, the error you are getting is in `CreateSampleDatabase` and not in `InitializeInstance`. What's happening is exactly what I was describing a couple of comments above. `InitializeInstance` is not being awaited so there's no guarantee that when `CreateSampleDatabase` executes the instance has actually finished creation. And that's why you are getting this error.
2. As for when you get the actual exception or an AggregateException, take a look at [Exception Handling (Task Parallel library)](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/exception-handling-task-parallel-library). Notice that because of the command runner we are calling Wait() on async methods (which is different from awaiting) and that's why we end up getting the AggregateException.

Reply:
Oh wow, thank you both for explaining what's going on here. Now I understand.

1. I've now made made `InitializeInstance` synchronous. This was because we couldn't call an async method within a lock.

2. Now that I realize where the exception was coming from, it makes sense why I was seeing an AggregateException instead of an RpcException. It was because the exception was coming from `CreateSampleDatabase` as @amanda-tarafa explained. Now that that issue is fixed, it throws and `RpcException` as expected here.